llms.txt

```diff
@@ -11289,6 +11290,780 @@ Doc-Content: https://docs.kluster.ai/tutorials/klusterai-api/image-analysis.ipyn
+     "text": [
+      "The result of 1337 multiplied by 42 is 56,154.\n"
+     ]
+    }
+   ],
+   "source": [
+    "import math\n",
+    "import re\n",
+    "\n",
+    "def execute_calculator(expression):\n",
+    "    # Simple calculator using eval() (note: never use this in production without proper validation)\n",
+    "    # In production, use a safer method for evaluation\n",
+    "    try:\n",
+    "        # Basic sanitization\n",
+    "        if not re.match(r'^[0-9+\\-*/().%\\s]+$', expression):\n",
+    "            return {\"error\": \"Invalid expression. Only basic arithmetic operations are allowed.\"}\n",
+    "        \n",
+    "        result = eval(expression)\n",
+    "        return {\"result\": result}\n",
+    "    except Exception as e:\n",
+    "        return {\"error\": str(e)}\n",
+    "\n",
+    "def process_tool_calls(response):\n",
+    "    message = response.choices[0].message\n",
+    "    \n",
+    "    # If there are no tool calls, return the message content\n",
+    "    if not message.tool_calls:\n",
+    "        return message.content\n",
+    "    \n",
+    "    # Process each tool call\n",
+    "    tool_results = []\n",
+    "    for tool_call in message.tool_calls:\n",
+    "        function_name = tool_call.function.name\n",
+    "        arguments = json.loads(tool_call.function.arguments)\n",
+    "        \n",
+    "        # Execute the appropriate function based on the tool call\n",
+    "        if function_name == \"calculator\":\n",
+    "            result = execute_calculator(arguments[\"expression\"])\n",
+    "            tool_results.append({\n",
+    "                \"tool_call_id\": tool_call.id,\n",
+    "                \"function_name\": function_name,\n",
+    "                \"result\": result\n",
+    "            })\n",
+    "    \n",
+    "    # Create a new message with the tool results\n",
+    "    messages = [\n",
+    "        {\"role\": \"user\", \"content\": \"What is 1337 multiplied by 42?\"},\n",
+    "        message.model_dump(),\n",
+    "    ]\n",
+    "    \n",
+    "    # Add the tool results\n",
+    "    for result in tool_results:\n",
+    "        messages.append({\n",
+    "            \"role\": \"tool\",\n",
+    "            \"tool_call_id\": result[\"tool_call_id\"],\n",
+    "            \"content\": json.dumps(result[\"result\"])\n",
+    "        })\n",
+    "    \n",
+    "    # Get the final response\n",
+    "    final_response = client.chat.completions.create(\n",
+    "        model=\"klusterai/Meta-Llama-3.1-8B-Instruct-Turbo\",\n",
+    "        messages=messages\n",
+    "    )\n",
+    "    \n",
+    "    return final_response.choices[0].message.content\n",
+    "\n",
+    "# Process the calculator response\n",
+    "final_answer = process_tool_calls(calculator_response)\n",
+    "print(final_answer)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Advanced Tool Calling Example - Web Search\n",
```

Comment:
Title capitalization

Reply:
Just realized this across the entire tutorial remember that kluster.ai does not use CamelCase

Reply:
This example is really cool! But I would divide this into multiple code snippets and then one that runs it all, with text in-between, it is a bit easier to follow